### PR TITLE
Feature 1858 obs quality

### LINKED ISSUE #1858: Add Point-Stat and Ensemble-Stat obs_quality_exc configuration option to specify which quality flags should be excluded
## Describe the New Feature ##
This feature was requested by the UK Met Office via [GitHub Discussions](https://github.com/dtcenter/METplus/discussions/1005).

Recommend the following changes:
- Rename the existing "obs_quality" config option for Point-Stat and Ensemble-Stat as "obs_quality_inc".
- Add a new config option for "obs_quality_exc" to specify which quality strings should be excluded.
- If "obs_quality_inc" is not defined, parse the old "obs_quality" config option and store use its value, for backwards compatibility.
- Be sure that these options are parsed separately for each verification tasks, just like the existing "obs_quality" option is.
- Be sure to update the documentation and unit tests to exercise this.

### Acceptance Testing ###
*List input data types and sources.*
*Describe tests required for new functionality.*

### Time Estimate ###
2 days.

### Sub-Issues ###
Consider breaking the new feature down into sub-issues.
No sub-issues needed.

### Relevant Deadlines ###
Needed for the operational use of METplus by the UK Met Office

### Funding Source ###
Met Office Support key: 2799991

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required: Seth
- [x] Select **scientist(s)** or **no scientist** required: no scientist required but request that the Met Office test the resulting beta release.

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [x] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
- [X] Define a METplus issue to handle the config option changes for obs_quality_inc and obs_quality_exc: dtcenter/METplus#1213

## New Feature Checklist ##
See the [METplus Workflow](https://dtcenter.github.io/METplus/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by sethlinden**
## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>

Ran several tests by hand for both point_stat and ensemble_stat:

Point-stat test for obs_quality_inc and obs_quality_exc:

setenv OBS_QUALITY_INC_SFC \"2\" ; setenv OBS_QUALITY_INC_UPA \"2\" ; setenv OBS_QUALITY_EXC_SFC \"9\" ; setenv OBS_QUALITY_EXC_UPA \"9\" ; setenv OUTPUT_PREFIX QTY_PB

./point_stat /d1/projects/MET/MET_test_data/unit_test/model_data/grib1/nam/nam_2012040900_F012.grib /d1/personal/linden/git/feature_1858_obs_quality/MET/test_output/pb2nc/gdas1.20120409.t12z.prepbufr.nc /d1/personal/linden/git/feature_1858_obs_quality/MET/test/config/PointStatConfig_qty_inc -outdir /d1/personal/linden/git/feature_1858_obs_quality/MET/test_output/point_stat -v 3

./point_stat /d1/projects/MET/MET_test_data/unit_test/model_data/grib1/nam/nam_2012040900_F012.grib /d1/personal/linden/git/feature_1858_obs_quality/MET/test_output/pb2nc/gdas1.20120409.t12z.prepbufr.nc /d1/personal/linden/git/feature_1858_obs_quality/MET/test/config/PointStatConfig_qty_exc -outdir /d1/personal/linden/git/feature_1858_obs_quality/MET/test_output/point_stat -v 3

Ensemble-stat test for obs_quality_inc and obs_quality_exc:

setenv OBS_QUALITY_INC \"V\" ; setenv OBS_QUALITY_EXC \"C\" ; setenv OUTPUT_PREFIX QTY_PB

./ensemble_stat 3 /d1/projects/MET/MET_test_data/unit_test/model_data/grib1/arw-fer-gep1/arw-fer-gep1_2012040900_F012.grib /d1/projects/MET/MET_test_data/unit_test/model_data/grib1/arw-sch-gep2/arw-sch-gep2_2012040900_F012.grib /d1/projects/MET/MET_test_data/unit_test/model_data/grib1/arw-tom-gep3/arw-tom-gep3_2012040900_F012.grib /d1/personal/linden/git/feature_1858_obs_quality/MET/test/config/EnsembleStatConfig_qty_inc -point_obs /d1/personal/linden/git/feature_1858_obs_quality/MET/test_output/madis2nc/metar_2012040912_F000.nc -outdir /d1/personal/linden/git/feature_1858_obs_quality/MET/test_output/ensemble_stat -v 3

./ensemble_stat 3 /d1/projects/MET/MET_test_data/unit_test/model_data/grib1/arw-fer-gep1/arw-fer-gep1_2012040900_F012.grib /d1/projects/MET/MET_test_data/unit_test/model_data/grib1/arw-sch-gep2/arw-sch-gep2_2012040900_F012.grib /d1/projects/MET/MET_test_data/unit_test/model_data/grib1/arw-tom-gep3/arw-tom-gep3_2012040900_F012.grib /d1/personal/linden/git/feature_1858_obs_quality/MET/test/config/EnsembleStatConfig_qty_exc -point_obs /d1/personal/linden/git/feature_1858_obs_quality/MET/test_output/madis2nc/metar_2012040912_F000.nc -outdir /d1/personal/linden/git/feature_1858_obs_quality/MET/test_output/ensemble_stat -v 3

I also ran all the unit tests by running "make test"

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

Run the unit test: unit_quality_filter.xml. Use the input and output files defined in the unit-test.

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes or No]**

Yes. I updated the documentation: point-stat.rst, ensemble-stat.rst, config_options.rst

- [x] Do these changes include sufficient testing updates? **[Yes or No]**

Yes. I updated unit_quality_filter.xml unit tests for both point_stat and ensemble_stat to run tests for obs_quality_inc and obs_quality_exc.

- [x] Will this PR result in changes to the test suite? **[Yes or No]**</br>

Yes/No. Changes added two more unit tests within unit_quality_filter.xml. Output files don't change though.

If **yes**, describe the new output and/or changes to the existing output:</br>

These updates don't change the output of point_stat or ensemble_stat.

- [ ] Please complete this pull request review by **[Fill in date]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [x] After submitting the PR, select **Linked issues** with the original issue number.
- [x] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [x] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

File: met/src/tools/core/ensemble_stat/ensemble_stat_conf_info.cc [hunk 1]
```diff
@@ -829,9 +829,24 @@ void EnsembleStatVxOpt::process_config(GrdFileType ftype, Dictionary &fdict,
    // Conf: sid_exc
    vx_pd.set_sid_exc_filt(parse_conf_sid_list(&odict, conf_key_sid_exc));
 
-   // Conf: obs_qty
-   vx_pd.set_obs_qty_filt(parse_conf_obs_qty(&odict));
-
+   // Conf: obs_qty_inc
+   // Check for old "obs_quality" field
+   StringArray obs_qty_sa = odict.lookup_string_array(conf_key_obs_qty);
```
Comment (JohnHalleyGotway) [inline met/src/tools/core/ensemble_stat/ensemble_stat_conf_info.cc]: All of this logic is already handled in parse_conf_obs_qty_inc() and should be removed from this spot, as you did in point_stat_conf_info.cc.

Comment (sethlinden) [reply met/src/tools/core/ensemble_stat/ensemble_stat_conf_info.cc]: Yes, that was an older version of the file. I modified it. I checked this morning that it was checked in. GitHub indicated it was already checked in. Let me try to re-push it up.

Comment (sethlinden) [reply met/src/tools/core/ensemble_stat/ensemble_stat_conf_info.cc]: I just pushed up the latest version. I don't know why it didn't recognize the change previously.

**Review by JohnHalleyGotway — COMMENTED** [review]
@sethlinden in an effort to avoid a proliferation of config files that are 99% the same, I'd recommend following the existing testing paradigm for the sid_inc_exc example:

- Delete EnsembleStatConfig_qty_inc, EnsembleStatConfig_qty_exc, PointStatConfig_qty_inc, and PointStatConfig_qyt_exc.
- Rename EnsembleStatConfig_qty to EnsembleStatConfig_qty_inc_exc.
- Rename PointStatConfig_qty to PointStatConfig_qty_inc_exc.
- Modify PointStatConfig_qyt_inc_exc and PointStatConfig_qyt_inc_exc to repeat the same fields already listed, but use obs_qty_inc in the first set and obs_qty_exc in the second set. Specify different descriptions (desc) to differentiate between the output. See PointStatConfig_sid_inc_exc for an example.
- Update the unit test XML file as needed to make calls using the updated EnsembleStatConfig_qty_inc_exc and PointStatConfig_qty_inc_exc files.

**Comment by JohnHalleyGotway** [comment]
Running "make test" on this branch results in this runtime error:
```
ERROR  : Dictionary::lookup_string_array() -> lookup failed for name "obs_quality"
```
I will debug this behavior this afternoon and push up some changes to address it.

**Review by JohnHalleyGotway — APPROVED** [review]
I approve of these changes.

Note that they DO change the result of the unit tests. So we'll need change the develop reference version soon. I tested this by running a full regression test in kiowa:/d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta3/feature_1858. Note that all changes listed in comp_dir_feature_1858_obs_quality_vs_develop.log are as expected.

1. 5 files with the prefix "ensemble_stat/ensemble_stat_QTY_MADIS_VGS" are now renamed to "ensemble_stat/ensemble_stat_QTY_INC_EXC_MADIS_VGS".
2. 3 files with the prefix "ensemble_stat/ensemble_stat_QTY_MADIS_VGS" are deleted since they are not needed to test the logic.
2. 4 files with the prefix "point_stat/point_stat_QTY_INC_EXC_PB" are renamed to "point_stat/point_stat_QTY_INC_EXC_PB".
3. 4 files with the prefix "point_stat/point_stat_QTY_INC_EXC_PB" are deleted since they are not needed to test the logic.

Carefully inspected the contents of point_stat_QTY_INC_EXC_PB_120000L_20120409_120000V_mpr.txt.
- Check ADPSFC ALL_QTY output (8 and 9):
```
grep ALL_QTY  point_stat_QTY_INC_EXC_PB_120000L_20120409_120000V_mpr.txt | grep ADPSFC | awk '{print $34}' | sort -u
8
9
```
- Confirmed that QTY_INC=8 and QTY_EXC=9 both result in only 8 in the output.

**Pull request merged**